### PR TITLE
Arbitrary permutation support in codegen integration

### PR DESCRIPTION
**Pull request by jjsjann123**
Extends permutation support to go beyond channels_last and contiguous to arbitrary permutation.

This is done by refactoring MemoryFormat from an enum class with fixed entries to a generic data structure holding permutation in std::vector<int>. This allows us to store multiple copies of a given tensor in different permutation.

A potential issues: Getting identical behavior as with eager could be tricky. e.g. batch_norm in eager supports arbitrary permutation, which would require us to update parsing our parsing rule. We have some simple binary tests as well as limited batch_norm test for now.

### CODE REVIEW

**Comment by jjsjann123** [comment]
This one should be good for review

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 1]
```diff
@@ -56,52 +56,188 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+
+  // default to non-permuted tensor
+  MemoryFormat() = default;
+
+  // stride_order is extracted from
+  //     `TensorType::stride_properties()::stride_index_`, it describes the
+  // index of axes from fastest to slowest.
+  // e.g. for rank 4 non-permuted tensor, stride_order would be {3, 2, 1, 0}
+  //      for rank 4 channels last tensor, stride_order would be {1, 3, 2, 0}
+  void setPermutation(std::vector<int> stride_order) {
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Could you note why this set as fastest to slowest given this is the opposite of how the strides appear?

Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Also, should the parameter `stride_order` be a `const` reference?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -56,52 +56,188 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+      permutation_ = 0;
+    }
+  }
+
+  // returns non-permuted format
+  static MemoryFormat Contiguous() {
+    return MemoryFormat();
+  }
+
+  bool hasPermutation() {
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: Nit: I don't know how worthwhile it is to go crazy with the nitpicking but it looks like all the methods from here on down could be marked as `const`.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -56,52 +56,188 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+  }
 
-// lower number has higher precedence, so order matters here and we currently
-// prioritize `ChannelsLast`
-enum class MemoryFormat { ChannelsLast = 0, Contiguous = 1 };
+  // returns transpose map to achieve permutation on non-permuted tensor
+  // note: used for aten::permute API
+  std::vector<int64_t> apply_vec() {
+    if (hasPermutation()) {
+      std::vector<int64_t> ret(permuted_order_.size());
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: NIT: I might pull the `ret` instantiation out and of the `if` to try and take advantage of return value optimization.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 4]
```diff
@@ -56,52 +56,188 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+    }
+    return {};
+  }
+
+  // returns transpose map to restore back to non-permuted tensor
+  // note: used for aten::permute API
+  std::vector<int64_t> restore_vec() {
+    if (hasPermutation()) {
+      int rank = permuted_order_.size();
+      std::vector<int64_t> ret(rank);
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: NIT: I might pull the `ret` instantiation out and of the `if` to try and take advantage of return value optimization.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 5]
```diff
@@ -56,52 +56,188 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
+};
+
+bool operator==(const MemoryFormat& a, const MemoryFormat& b) {
+  return a.permutation_ == b.permutation_;
+};
+
+typedef std::map<MemoryFormat, CgValue, MemoryCompare> MemoryFormatMap;
 
-// return format with higher precedence, this is used in folding expression
 MemoryFormat operator+(const MemoryFormat& a, const MemoryFormat& b) {
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp:248]: Why would you want to add two MemoryFormat structs?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 6]
```diff
@@ -56,52 +56,188 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
...
-  return a <= b ? a : b;
+  // Note: TensorIterator logic uses first input to dominate output MemoryFormat
+  // so instead of `a.permutation_ >= b.permutation_ ? a : b;`, we use:
+  return a;
 };
 
 class ValueHolder {
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/parser.cpp:261]: Comment on purpose since it is widely used?

**Review by kevinstephano — APPROVED** [review]
Perhaps there a few things to look at that are nitpicks and a couple of requests for a some comments on purpose.  I thought maybe ValueHolder needed a comment on purpose even though it wasn't changed in this PR.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Thanks for guarding the quality of code. It's well appreciated.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 5 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:248]: We don't really need this any more and could have used a properly named function. Back then when I was using binary fold to merge memory format, I overloaded add operator.

Right now it's still used when we want to combine two tensor with different memory format.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: That's a choice made by PE guys when they profile tensors. I can put a link to the code comment

**Comment by jjsjann123** [comment]
Local tests passing except for non-divisible split in layer_norm. I'm merging this PR

**Pull request merged**